tests/installation/test_authenticator.py

```diff
@@ -201,3 +202,50 @@ def test_authenticator_uses_env_provided_credentials(
     request = http.last_request()
 
     assert "Basic YmFyOmJheg==" == request.headers["Authorization"]
+
+
+def test_authenticator_uses_certs_from_config_if_not_provided(
+    config, mock_remote, http, mocker
+):
+    config.merge(
+        {
+            "repositories": {"foo": {"url": "https://foo.bar/simple/"}},
+            "http-basic": {"foo": {"username": "bar", "password": "baz"}},
+            "certificates": {
+                "foo": {"cert": "/path/to/cert", "client-cert": "/path/to/client-cert"}
+            },
+        }
+    )
+
+    authenticator = Authenticator(config, NullIO())
+    session_send = mocker.patch.object(authenticator.session, "send")
+    authenticator.request("get", "https://foo.bar/files/foo-0.1.0.tar.gz")
+    kwargs = session_send.call_args[1]
+    assert pathlib.Path(kwargs["verify"]) == pathlib.Path("/path/to/cert")
+    assert pathlib.Path(kwargs["cert"]) == pathlib.Path("/path/to/client-cert")
+
+
+def test_authenticator_uses_provided_certs_instead_of_config_certs(
+    config, mock_remote, http, mocker
+):
+    config.merge(
+        {
+            "repositories": {"foo": {"url": "https://foo.bar/simple/"}},
+            "http-basic": {"foo": {"username": "bar", "password": "baz"}},
+            "certificates": {
+                "foo": {"cert": "/path/to/cert", "client-cert": "/path/to/client-cert"}
+            },
+        }
+    )
+
+    authenticator = Authenticator(config, NullIO())
+    session_send = mocker.patch.object(authenticator.session, "send")
+    authenticator.request(
+        "get",
+        "https://foo.bar/files/foo-0.1.0.tar.gz",
+        verify="/path/to/provided/cert",
+        cert="/path/to/provided/client-cert",
+    )
+    kwargs = session_send.call_args[1]
+    assert pathlib.Path(kwargs["verify"]) == pathlib.Path("/path/to/provided/cert")
+    assert pathlib.Path(kwargs["cert"]) == pathlib.Path("/path/to/provided/client-cert")
```

Comment:
```suggestion
@pytest.mark.parametrize("cert,client_cert", [
    (None, None),
    ("/path/to/provided/cert", "/path/to/provided/client-cert")
])
def test_authenticator_uses_certs_from_config_if_not_provided(
    cert, client_cert, config, mock_remote, http, mocker
):
    configured_cert = "/path/to/cert"
    configured_client_cert = "/path/to/client-cert"
    config.merge(
        {
            "repositories": {"foo": {"url": "https://foo.bar/simple/"}},
            "http-basic": {"foo": {"username": "bar", "password": "baz"}},
            "certificates": {
                "foo": {"cert": configured_cert, "client-cert": configured_client_cert}
            },
        }
    )

    authenticator = Authenticator(config, NullIO())
    session_send = mocker.patch.object(authenticator.session, "send")
    authenticator.request("get", "https://foo.bar/files/foo-0.1.0.tar.gz", verify=cert, cert=client_cert)
    kwargs = session_send.call_args[1]
    
    assert Path(kwargs["verify"]) == Path(cert or configured_cert)
    assert Path(kwargs["cert"]) == Path(client_cert or configured_client_cert)
```
Nothing wroing with the implementation, just wanted to avoid some duplicated code.

Reply:
I think it makes sense to have the tests separated since the test name clearly defines the conditions under test. In this suggestion, the parameterization makes the test cases less clear.

Reply:
I am not wedded to parametrised tests. But I am much more keen on adding to the already large sets of duplicated code in our test suites. It is becoming harder to maintain. Additionally you can add a comment next to the parameters if further clarity is required. Personally, I do think the parameters are clear enough.

Reply:
OK. Will update

Reply:
This has been updated as requested